internal/js/reader.go

```diff
@@ -0,0 +1,62 @@
+package js
+
+import (
+	"context"
+	"io"
+)
+
+var _ io.Reader = (*ChannelReader)(nil)
+
+// ChannelReader is a reader that reads from a channel of bytes.
+type ChannelReader struct {
+	buff     []byte
+	source   <-chan []byte
+	cancelFn context.CancelFunc
+}
+
+// NewChannelReader creates a new ChannelReader from a channel.
+//
+// Second argument is optional function that will be called when `Close` method is called.
+func NewChannelReader(source <-chan []byte, cancelFn context.CancelFunc) *ChannelReader {
+	return &ChannelReader{
+		source: source,
```

Comment:
You might have missed this one:

```suggestion
		source: source,
		cancelFn: cancelFn,
```